cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/operation/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,145 @@
+  private final Lock lock;
+  private final ExecutorService executor;
+  private LongRunningOperation<T> currentOperation;
+
+  /**
+   * Default constructor.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
+      OperationStatePublisher statePublisher) {
+    super(operations);
+    this.statePublisher = statePublisher;
+    this.lock = new ReentrantLock();
+    this.executor = Executors.newSingleThreadExecutor(
+        Threads.createDaemonThreadFactory("operation-runner-%d")
+    );
+  }
+
+  // TODO(samik) It is currently rudimentary implementation to unblock the state management
+  //  Handle states, publish failures and heartbeat messages
+  @Override
+  public void run(LongRunningOperationRequest<T> request)
+      throws OperationTypeNotSupportedException, InvalidRequestTypeException {
+    lock.lock();
+    try {
+      if (currentOperation != null) {
+        throw new IllegalStateException("Operation already running");
+      }
+      currentOperation = getOperation(request);
+      executor.execute(() -> {
+        runInternal(request);
+      });
+      statePublisher.publishRunning();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void runInternal(LongRunningOperationRequest<T> request) {
+    try {
+      ListenableFuture<OperationError> operationTask = currentOperation.run(request.getOperationRequest(), this::sendMetaUpdateMessage);
+      Futures.addCallback(
+          operationTask,
+          new FutureCallback<OperationError>() {
+            @Override
+            public void onSuccess(@Nullable OperationError result) {
+              if (result != null) {
+                statePublisher.publishFailed(result);
+              } else {
+                statePublisher.publishSuccess();
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              statePublisher.publishFailed(getOperationErrorFromThrowable(request, t));
+            }
+          }
+      );
+    } catch (Exception e) {
+      statePublisher.publishFailed(getOperationErrorFromThrowable(request, e));
+    }
+  }
+
+  // TODO(samik): Implement graceful shutdown
+  @Override
+  public void stop(long timeout, TimeUnit timeoutUnit) throws Exception {
+    close();
+  }
+
+  private void sendMetaUpdateMessage(OperationMeta meta) {
+    statePublisher.publishMetaUpdate(meta);
+  }
+
+  private OperationError getOperationErrorFromThrowable(LongRunningOperationRequest<T> request,
+      Throwable t) {
+    OperationRunId runId = request.getRunId();
+    LOG.error("Operation {} of namespace {} failed: {}", runId.getRun(), runId.getParent(),
+        t.getMessage());
+    return new OperationError(t.getMessage(), Collections.emptyList());
+  }
+
+  @Override
+  public void close() throws Exception {
+    lock.lock();
+    try {
+      if (currentOperation == null) {
+        throw new IllegalStateException("No operation is running");
```

Comment:
If throwing here, then the shutdownNow and publishStopped won't get call.

Reply:
If the current operation is null, there would be no running thread i.e no run has called for this runner. In that case I don't think we should run `shutdownNow` or `publishStopped`.